aten/src/ATen/native/cudnn/RNN.cpp

```diff
@@ -0,0 +1,896 @@
+}
+
+}} // namespace at::native
+
+#else // AT_CUDNN_ENABLED()
+
+#include <ATen/cudnn/cudnn-wrapper.h>
+#include <ATen/cudnn/Descriptors.h>
+#include <ATen/cudnn/Types.h>
+#include <ATen/cudnn/Utils.h>
+
+namespace at { namespace native {
+
+namespace {
+  // DropoutDescriptor
+
+  struct DropoutDescriptorParams {
+    bool train;
+    double dropout;
+    Tensor dropout_state;
+    DropoutDescriptorParams() {}
+    void set(bool train_, double dropout_, Tensor dropout_state_) {
+      train = train_;
+      dropout = dropout_;
+      dropout_state = dropout_state_;
+    }
+    DropoutDescriptor descriptor(cudnnHandle_t handle) const {
+      // NB: dropout_seed passed dummy 0, because it isn't actually used
+      // when dropout_state is defined.
+      auto dropout_p = train ? dropout : 0;
+      DropoutDescriptor dropout_desc;
+      dropout_desc.set(handle, dropout_p, dropout_state, 0);
+      return dropout_desc;
+    }
+  };
+
+  // RNNDescriptor
+
+  struct RNNDescriptorParams {
+    int64_t hidden_size;
+    int64_t num_layers;
+    cudnnDirectionMode_t bidirectional;
+    cudnnRNNMode_t mode;
+    cudnnDataType_t datatype;
+
+    cudnnRNNInputMode_t input_mode = CUDNN_LINEAR_INPUT;
+
+    int64_t num_directions() const {
+      return bidirectional ? 2 : 1;
+    }
+
+    void set_mode(int64_t fn_mode) {
+      switch (fn_mode) {
+        case CUDNN_RNN_RELU:
+          mode = CUDNN_RNN_RELU;
+          break;
+        case CUDNN_RNN_TANH:
+          mode = CUDNN_RNN_TANH;
+          break;
+        case CUDNN_LSTM:
+          mode = CUDNN_LSTM;
+          break;
+        case CUDNN_GRU:
+          mode = CUDNN_GRU;
+          break;
+        default:
+          throw std::runtime_error("unrecognized mode"); // TODO
+      }
+    }
+
+    void set_bidirectional(bool fn_bidirectional) {
+      bidirectional = fn_bidirectional ? CUDNN_BIDIRECTIONAL : CUDNN_UNIDIRECTIONAL;
+    }
+
+    RNNDescriptor descriptor(cudnnHandle_t handle, DropoutDescriptor&& dropout_desc) const {
+      RNNDescriptor rnn_desc;
+      rnn_desc.set(handle, hidden_size, num_layers, std::move(dropout_desc), input_mode, bidirectional, mode, datatype);
+      return rnn_desc;
+    }
+
+    RNNDescriptor descriptor(cudnnHandle_t handle) const {
+      DropoutDescriptor dropout_desc;
+      dropout_desc.set(handle, 0, {}, 0);
+      return descriptor(handle, std::move(dropout_desc));
+    }
+  };
+
+  // TensorDescriptor list
+
+  std::vector<TensorDescriptor> rnn_descriptor_sequence(const Tensor& tensor, IntList batch_sizes) {
+    std::vector<TensorDescriptor> descriptors(batch_sizes.size());
+    size_t i = 0;
+    for (auto batch_size : batch_sizes) {
+      // NB: The narrow is solely to adjust the batch size; to do it
+      // accurately we would have to adjust the start index as well,
+      // but the pointer location isn't actually used so we can skip it.
+      // NB: cuDNN RNN API has an undocumented requirement that all
+      // tensors have dimension 5.
```

Comment:
though current python bindings do create 5d descriptors, there's no such requirement. In face, RNN sample that is distributed with cudnn creates 3d descriptors (2d would be enough, but cudnn does not support 2d, don't ask).

Reply:
Yeah, it looks like we only need to pad out to 3; agreed!